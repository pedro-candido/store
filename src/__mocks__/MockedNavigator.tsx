import React from 'react';

import { NavigationContainer } from '@react-navigation/native';
import { createStackNavigator } from '@react-navigation/stack';

const Stack = createStackNavigator();

type Params = {
  component: React.ComponentType<any>;
  [key: string]: any;
};

const MockedNavigator = ({ component, params = {} }: Params) => {
  return (
    <NavigationContainer>
      <Stack.Navigator>
        <Stack.Screen name="MockedScreen" component={component} initialParams={params} />
      </Stack.Navigator>
    </NavigationContainer>
  );
};

export default MockedNavigator;
