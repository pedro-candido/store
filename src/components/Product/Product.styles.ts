import styled from 'styled-components/native';

import { MaterialCommunityIcons } from '@expo/vector-icons';

export const Container = styled.View`
  justify-content: center;
  align-items: center;
  width: 100%;
  height: 100%;
  padding: 10px;
`;

export const Picture = styled.Image`
  width: 150px;
  height: 150px;
  margin-bottom: 20px;
  border-radius: 5px;
`;

export const PriceContainer = styled.View`
  flex-direction: row;
  justify-content: flex-start;
`;

export const RatingContainer = styled.View`
  flex-direction: row;
  justify-content: flex-start;
  margin-top: 5%;
  align-items: flex-start;
  margin-bottom: auto;
  align-content: flex-start;
`;

export const Wrapper = styled.View`
  flex-direction: row;
  justify-content: space-between;
  align-content: center;
  align-items: flex-end;
`;

export const StarIcon = styled(MaterialCommunityIcons).attrs({
  color: '#fdc509',
  size: 14,
})``;
