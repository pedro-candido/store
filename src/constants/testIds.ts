const testIds = {
  addToCartButton: 'add-to-cart-button',
  removeFromCartButton: 'remove-from-cart-button',
  productTestId: 'product-',
  backButton: 'back-button',
  cartButton: 'cart-button',
};

export { testIds };
