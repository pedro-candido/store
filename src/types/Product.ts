export type TProduct = {
  id: number;
  title: string;
  price: number;
  description: string;
  category: string;
  image: string;
  rating: {
    rate: number;
    count: number;
  };
};
