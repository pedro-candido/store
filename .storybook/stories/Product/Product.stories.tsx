import { View } from 'react-native';
import type { Meta, StoryObj } from '@storybook/react';
import Product from '@/components/Product';
import { products } from '@/__mocks__/productsMock';

const ProductMeta: Meta<typeof Product> = {
  title: 'Product',
  component: Product,
  args: {
    children: 'Hello world',
  },
  decorators: [
    (Story) => (
      <View style={{ alignItems: 'center', justifyContent: 'center', flex: 1, width: '50%' }}>
        <Story />
      </View>
    ),
  ],
};

export default ProductMeta;

export const Default: StoryObj<typeof Product> = {
  args: {
    ...products[0],
  },
};
