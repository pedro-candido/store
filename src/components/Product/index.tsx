import { View } from 'react-native';

import Button from '@/components/Button';
import { TProduct } from '@/types/Product';
import Label from '@components/Label';
import {
  Container,
  Picture,
  PriceContainer,
  RatingContainer,
  StarIcon,
  Wrapper,
} from '@components/Product/Product.styles';
import { testIds } from '@constants/testIds';
import { useProducts } from '@hooks/useProducts';
import { useStore } from '@store/globalStore';
import { convertNumberToDollar } from '@utils/formatter';

const { productTestId } = testIds;

function Product({ id, title, price, description, image, rating }: TProduct) {
  const { addProduct, productsChosen, removeProduct } = useStore();
  const { hasProductOnCart } = useProducts();

  const isProductAlreadyAddedToCard = hasProductOnCart(id, productsChosen);

  function handlePressPrimary() {
    addProduct({
      productId: id,
      productTestId: `${productTestId}${id}`,
      image,
      price,
      title,
    });
  }

  function handlePressSecondary() {
    removeProduct(id);
  }

  return (
    <Container>
      <View>
        <Picture
          source={{
            uri: image,
          }}
        />
      </View>
      <View>
        <View>
          <Label numberOfLines={2}>
            {title}. {description}
          </Label>
          <Wrapper>
            <RatingContainer>
              <StarIcon name={'star'} />
              <Label>
                {rating.rate}({rating.count})
              </Label>
            </RatingContainer>
            <PriceContainer>
              <Label>{convertNumberToDollar(price)}</Label>
            </PriceContainer>
          </Wrapper>
          <View>
            {!isProductAlreadyAddedToCard ? (
              <Button
                testID={`add-${productTestId}${id}`}
                iconName={'cart'}
                text={'Add to Cart'}
                onPress={handlePressPrimary}
              />
            ) : (
              <Button
                testID={`remove-${productTestId}${id}`}
                type={'secondary'}
                iconName={'cart-remove'}
                text={'Remove'}
                onPress={handlePressSecondary}
              />
            )}
          </View>
        </View>
      </View>
    </Container>
  );
}

export default Product;
