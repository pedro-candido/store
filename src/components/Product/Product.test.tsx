import React from 'react';

import { fireEvent, render } from '@testing-library/react-native';

import { products } from '@/__mocks__/productsMock';
import { testIds } from '@constants/testIds';

import Product from './index';

const item = products[0];

const mockedRemoveProduct = jest.fn();
const mockedAddProduct = jest.fn();
const mockedHasProduct = jest.fn();

const { productTestId } = testIds;

jest.mock('@store/globalStore', () => ({
  useStore: () => ({
    removeProduct: mockedRemoveProduct,
    addProduct: mockedAddProduct,
  }),
}));

jest.mock('@hooks/useProducts/useProducts', () => ({
  useProducts: () => ({
    hasProductOnCart: mockedHasProduct,
  }),
}));

describe('Product', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });
  it('should match snapshot', () => {
    const testRenderer = render(<Product {...item} />);

    expect(testRenderer.toJSON).toMatchSnapshot();
  });
  it('should press on add button', () => {
    const { getByTestId } = render(<Product {...item} />);
    const addButton = getByTestId(`add-${productTestId}${item.id}`);
    fireEvent.press(addButton);

    expect(mockedAddProduct).toHaveBeenCalled();
  });
});

describe('Product with remove', () => {
  beforeEach(() => {
    mockedHasProduct.mockImplementation(() => true);
  });

  it('should press on remove button', () => {
    const { getByTestId } = render(<Product {...item} />);

    const button = getByTestId(`remove-${productTestId}${item.id}`);

    fireEvent.press(button);

    expect(mockedRemoveProduct).toHaveBeenCalled();
  });
});
