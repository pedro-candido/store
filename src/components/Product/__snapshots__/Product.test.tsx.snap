// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product should match snapshot 1`] = `[Function]`;
